lib/components/mixins/tapable.dart

```diff
@@ -1,74 +1,106 @@
       _currentPointerId = pointerId;
-      onTapDown(details);
+      return onTapDown(details);
     }
+    return true;
   }
 
-  void handleTapUp(Rect rect, int pointerId, TapUpDetails details) {
+  bool handleTapUp(int pointerId, TapUpDetails details) {
     if (_checkPointerId(pointerId) &&
-        checkTapOverlap(rect, details.localPosition)) {
+        checkOverlap(details.localPosition.toVector2())) {
       _currentPointerId = null;
-      onTapUp(details);
+      return onTapUp(details);
     }
+    return true;
   }
 
-  void handleTapCancel(int pointerId) {
+  bool handleTapCancel(int pointerId) {
     if (_checkPointerId(pointerId)) {
       _currentPointerId = null;
-      onTapCancel();
+      return onTapCancel();
     }
+    return true;
   }
 }
 
 mixin HasTapableComponents on BaseGame {
-  Iterable<PositionComponent> _positionComponents() {
-    return components.whereType<PositionComponent>();
-  }
-
   @mustCallSuper
   void onTapCancel(int pointerId) {
-    _positionComponents().forEach((c) {
-      c.propagateToChildren<Tapable>(
-        (child, _) => child.handleTapCancel(pointerId),
-      );
-    });
+    for (Component c in components.toList().reversed) {
+      bool shouldContinue = true;
+      if (c is BaseComponent) {
+        shouldContinue = c.propagateToChildren<Tapable>(
+          (child) => child.handleTapCancel(pointerId),
+        );
+      }
+      if (c is Tapable && shouldContinue) {
+        shouldContinue = c.handleTapCancel(pointerId);
+      }
+      if (!shouldContinue) {
+        break;
+      }
+    }
   }
 
   @mustCallSuper
   void onTapDown(int pointerId, TapDownDetails details) {
-    _positionComponents().forEach((c) {
-      c.propagateToChildren<Tapable>(
-        (child, rect) => child.handleTapDown(rect, pointerId, details),
-      );
-    });
+    for (Component c in components.toList().reversed) {
+      bool shouldContinue = true;
+      if (c is BaseComponent) {
+        shouldContinue = c.propagateToChildren<Tapable>(
+          (child) => child.handleTapDown(pointerId, details),
+        );
+      }
+      if (c is Tapable && shouldContinue) {
+        shouldContinue = c.handleTapDown(pointerId, details);
+      }
+      if (!shouldContinue) {
+        break;
+      }
+    }
   }
 
   @mustCallSuper
   void onTapUp(int pointerId, TapUpDetails details) {
-    _positionComponents().forEach((c) {
-      c.propagateToChildren<Tapable>(
-        (child, rect) => child.handleTapUp(rect, pointerId, details),
-      );
-    });
+    for (Component c in components.toList().reversed) {
+      bool shouldContinue = true;
+      if (c is BaseComponent) {
+        shouldContinue = c.propagateToChildren<Tapable>(
+          (child) => child.handleTapUp(pointerId, details),
+        );
+      }
+      if (c is Tapable && shouldContinue) {
+        shouldContinue = c.handleTapUp(pointerId, details);
+      }
+      if (!shouldContinue) {
+        break;
+      }
+    }
```

Comment:
I think these three methods can be only one right? they are almost identical. just make a genericHandler((c) => c.handleTapUp(...))

Reply:
Good idea, much cleaner solution now with less code duplication